### DISCUSSION TITLE
How can I run the model for a set of scenarios?

### DISCUSSION THREAD

**Discussion by aganter** [ZEN-universe/ZEN-garden#166] [Installation & Setup]
# Conducting scenario analysis
You can define your set of scenarios in the scenarios dict. The scenarios dict is defined in _scenarios.py_ and stored in the dataset folder similar to _system.py_. If you want to conduct your analysis for multiple scenarios, you have to set ```system["conductScenarioAnalysis"] = True```. The default in _defaultConfig.py_ is False.

## Defining scenarios
Each scenario has a unique name which serves as the key in the scenario dictionary. The corresponding value is a dictionary contains the elements (keys) and parameters (values), that change with respect to the base configuration of the model. 
```Python
scenarios[scenarioName] = {element1: [paramName1, ..., paramNameN], 
                           ..., 
                           elementN: [paramName1, ..., paramNameN]}
``` 

For example, if we want to change the carbonEmissionsBudget, we would write the following.
```Python
scenarios["newCarbonBudget"] = {"EnergySystem": ["carbonEmissionsBudget"]}
```

Between each run, the base configuration is restored. Thereafter, the thscenario-dependent parameters are updated. 

It is possible to modify time-dependent parameters (intra- and inter-yearly). In the case of intra-yearly time-dependencies, the timeSeriesAggregation is conducted again. However, please note that this is **not supported** for parameters that are **constant** in the base configuration! 

**Note:** If the parameter is not an attribute, but stored together with other parameters such as _priceCarrier"_, a tuple is used to pass the information on which parameter should be modified. For example, if the import price of the carrier electricity is modified, we would write the following:
```Python
scenarios["newImportPriceElectricity"] = {"electricity": [("priceCarrier", "importPriceCarrier")]}
```

## Storing scenario dependent values
The scenario dependent parameter values are stored in the same folder as the file of the base configuration. The file of the scenario dependent parameter values has the same format as the original file, where the original filename is extended by the unique scenario name. For example, if the scenario name is _scenario1_ the filename would be _originalFileName_scenario1.csv_

## Changing the base scenario
Per default the base configuration is restored between each run. This can be changed by calling from your compile
```
Python optimization_setup.setBaseConfiguration(scenario,elements)
``` 
where _scenario_ describes the name of the scenario that you want to use as your base scenario, and _elements_ is a dictionary containing the elements that are changed. The dictionary _elements_ is continuously updated to ensure that the base configuration can be restored between each run. Please note that independently of how you name your base scenario, per default the original csv-files will be used! Hence, the base scenario is the only case, where the scenario name is not appended to the filename.

# Troubleshooting:
- make sure ```system["conductScenarioAnalysis"] = True```
- make sure the scenario dict is updated in your config file:

```Python
## overwrite default system and scenario dictionaries 
if system["conductScenarioAnalysis"]:     
assert os.path.exists(f"data/{analysis['dataset']}/scenarios.py"), f"scenarios.py is missing from selected dataset '{analysis['dataset']}" 
spec    = importlib.util.spec_from_file_location("module", f"data/{analysis['dataset']}/scenarios.py")   
module  = importlib.util.module_from_spec(spec)     
spec.loader.exec_module(module)     
scenarios  = module.scenarios
```

**Comment by jacob-mannhardt** [comment]
@aganter is this outdated?
